engine/access/rest/blocks.go

```diff
@@ -0,0 +1,260 @@
+	if statusErr != nil {
+		return nil, statusErr
+	}
+
+	// lookup execution result
+	// (even if not specified as expandable, since we need the execution result ID to generate its expandable link)
+	executionResult, err := backend.GetExecutionResultForBlockID(req.Context(), blk.ID())
+	if err != nil {
+		return nil, err
+	}
+
+	return blockResponse(blk, executionResult, link, req.expandFields)
+}
+
+func idLookupError(id *flow.Identifier, entityType string, err error) StatusError {
+	msg := fmt.Sprintf("error looking up %s with ID %s", entityType, id.String())
+	// if error has GRPC code NotFound, then return HTTP NotFound error
+	if status.Code(err) == codes.NotFound {
+		return NewNotFoundError(msg, err)
+	}
+	return NewRestError(http.StatusInternalServerError, msg, err)
+}
+
+// todo(sideninja) refactor and merge
+func heightLookupError(height uint64, entityType string, err error) StatusError {
+	msg := fmt.Sprintf("error looking up %s at height %d", entityType, height)
+	// if error has GRPC code NotFound, then return HTTP NotFound error
+	if status.Code(err) == codes.NotFound {
+		return NewNotFoundError(msg, err)
+	}
+	return NewRestError(http.StatusInternalServerError, msg, err)
+}
+
+// blockProvider is a layer of abstraction on top of the backend access.API and provides a uniform way to
+// lookup a block or a block header either by ID or by height
+type blockProvider struct {
+	id      *flow.Identifier
+	height  uint64
+	final   bool
+	sealed  bool
+	backend access.API
+}
+
+type blockProviderOption func(blkProvider *blockProvider)
+
+func forID(id *flow.Identifier) blockProviderOption {
+	return func(blkProvider *blockProvider) {
+		blkProvider.id = id
+	}
+}
+func forHeight(height uint64) blockProviderOption {
+	return func(blkProvider *blockProvider) {
+		blkProvider.height = height
+	}
+}
+
+func forFinalized(queryParam string) blockProviderOption {
+	return func(blkProvider *blockProvider) {
+		switch queryParam {
+		case sealedHeightQueryParam:
+			blkProvider.sealed = true
+			fallthrough
+		case finalHeightQueryParam:
+			blkProvider.final = true
+		}
+	}
+}
+
+func NewBlockProvider(backend access.API, options ...blockProviderOption) *blockProvider {
+	blkProvider := &blockProvider{
+		backend: backend,
+	}
+
+	for _, o := range options {
+		o(blkProvider)
+	}
+	return blkProvider
+}
+
+func (blkProvider *blockProvider) getBlock(ctx context.Context) (*flow.Block, StatusError) {
+	if blkProvider.id != nil {
+		blk, err := blkProvider.backend.GetBlockByID(ctx, *blkProvider.id)
+		if err != nil {
+			return nil, idLookupError(blkProvider.id, "block", err)
+		}
+		return blk, nil
+	}
+
+	if blkProvider.final {
+		blk, err := blkProvider.backend.GetLatestBlock(ctx, blkProvider.sealed)
+		if err != nil {
+			// cannot be a 'not found' error since final and sealed block should always be found
+			return nil, NewRestError(http.StatusInternalServerError, "block lookup failed", err)
+		}
+		return blk, nil
+	}
```

Comment:
nit: I know this makes the code more consise, but I find this logic a bit unclear / confusing. 

Suggestion: could we rename the `final` field to `latest`? This would make things more clear imo